flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TimeGauge.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// ----------------------------------------------------------------------------
+//  This class is largely adapted from "com.google.common.base.Preconditions",
+//  which is part of the "Guava" library.
+//
+//  Because of frequent issues with dependency conflicts, this class was
+//  added to the Flink code base to reduce dependency on Guava.
+// ----------------------------------------------------------------------------
+
+package org.apache.flink.runtime.metrics.groups;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.View;
+
+/**
+ * {@link TimeGauge} encapsulate logic of measuring time rates (like idleTime ms/s). In particularly
+ * it takes care of the case, when single measurement (as defined by {@link #markStart()} and {@link
+ * #markEnd()}), crosses the boundary of single {@link #update()} interval. Thanks to the {@link
+ * #markStart()} {@link TimeGauge} can correctly calculate next value in {@link #update()}, even if
+ * this measurement hasn't not ended ({@link #markEnd()}).
+ */
+public class TimeGauge implements Gauge<Long>, View {
+    private long previousCount;
+    private long currentCount;
+    private long lastDifference;
+    private long currentMeasurementStart;
+
+    public synchronized void markStart() {
+        if (currentMeasurementStart == 0) {
+            currentMeasurementStart = System.currentTimeMillis();
```

Comment:
What is the scenario for that to happen?

`BufferWritingResultPartition` call happens only when processing input, so it's outside of the [`markStart`, `markEnd`] calls from the the `StreamTask`?

Reply:
I think you are right.
I was concerned about `MailboxProcessor` updating `idleTime` but I see it's removed so it should be fine.

Reply:
I agree this is a bit error-prone, but I think adding separate `TimerGauge` per each place where we are starting a measurement is probably not the way to go. It could quickly multiply the number of declared fields/gauges in `TaskIOMetricGroup` to facilitate every call place.

If this will prove to be a real issue, we could solve this issue for example by bumping a counter in each `markStart()` call. `markEnd` would decrease that counter, and we would measure the time between first `markStart` and last `markEnd` call. But I think that's over engineering at the moment?

Reply:
There are only two such places ATM and I don't think adding a (final) object per partition creates a visible overhead.
But I agree we can defer this discussion.